raft: fix the shutdown phase being stuck

emdroid opened the pull request:
Some of the calls inside the `raft_group0_client::start_operation()` method were missing the abort source parameter. This caused the repair test to be stuck in the shutdown phase - the abort source has been triggered, but the operations were not checking it.

This was in particular the case of operations that try to take the ownership of the raft group semaphore (`get_units(semaphore)`) - these waits should be cancelled when the abort source is triggered.

This should fix the following tests that were failing in some percentage of dtest runs (about 1-3 of 100):
* TestRepairAdditional::test_repair_kill_1
* TestRepairAdditional::test_repair_kill_3

Fixes scylladb/scylladb#19223

emdroid commented:
Tested with running both tests that had the failures occurring for 500 repetitions (usually failed 1-3 of 100 before), all now green:
* [TestRepairAdditional::test_repair_kill_1](https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2403/)
* [TestRepairAdditional::test_repair_kill_3](https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2403/)

kbr-scylla commented on service/raft/raft_group0_client.cc:
```diff
@@ -251,12 +251,22 @@ future<group0_guard> raft_group0_client::start_operation(seastar::abort_source*
         case group0_upgrade_state::use_post_raft_procedures: {
-            auto operation_holder = co_await get_units(_operation_mutex, 1);
+            auto operation_holder = co_await std::invoke([this, as] {
+                if (!as) {
```
How much effort would it take to refactor `start_operation` to take `abort_source&` instead of `abort_source*`? (I wouldn't be surprised if all callsites already pass an `abort_source`)

emdroid replied:
Yeah, will have a look maybe it is the case (so we could actually have just one call)

emdroid replied:
See the comment - there is a place where we probably need to keep the pointer, but we can use work-around of creating a dummy abort source in that case ... and still looks worth the change as it simplifies the code significantly.

kbr-scylla commented on service/raft/raft_group0_client.cc:
```diff
@@ -251,12 +251,22 @@ future<group0_guard> raft_group0_client::start_operation(seastar::abort_source*
+                }
+                return get_units(_operation_mutex, 1, *as);
+            });
             co_await _raft_gr.group0_with_timeouts().read_barrier(as, timeout);
```
We should eventually also fix the semaphore waits to respect the passed-in `raft_timeout` as well, but that's out of scope of this PR and not urgent.

Please open a follow-up issue for that.

emdroid replied:
Created the #19867 for that.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10525/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10525/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10739/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2717/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10525/)
#### Build Details:
- Duration: 2 hr 43 min
- Builder: spider2.cloudius-systems.com

emdroid commented:
New version:
* rebased
* passing the abort source to the `raft_group0_client::start_operation()` by reference

Passing the abort source by ref simplifies to code heavily; there is a place though where we still get the abort source by pointer (which can be null) in "auth/common.cc" and which doesn't seem to be easy to change (ripples down into `add_entry` etc.), so created a work-around there using the `composite_abort_source` (which remains dummy - empty if the abort source is a null) - still worth the change IMHO as the rest of the code is simplified quite a lot.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10544/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10544/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/group0_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10544/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2727/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10748/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10544/)
#### Build Details:
- Duration: 2 hr 51 min
- Builder: spider7.cloudius-systems.com

emdroid commented:
rebased on master

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10554/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10554/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/group0_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10554/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2744/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10764/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10554/)
#### Build Details:
- Duration: 2 hr 51 min
- Builder: spider2.cloudius-systems.com

bhalevy reviewed (approved):
LGTM

bhalevy commented:
@kbr-scylla are you ok with the PR?
Can it be queued?

kbr-scylla commented on auth/common.cc:
```diff
@@ -190,7 +191,11 @@ future<> announce_mutations(
+    utils::composite_abort_source composite_as;
+    if (as) {
+        composite_as.add(*as);
+    }
```
Please dig deeper (or rather, go opposite of deeper? up the callstack) to investigate further if this can be avoided.
If not, we need to leave a FIXME and open an issue (but preferably should be fixed right here.)

By masking the problem like this we can potentially shoot ourselves in our feet, with another shutdown hang in the future, if this `announce_mutations` call is started concurrently.

emdroid replied:
Done ... I was stopping here because the change ripples through even into the `add_entry()`, but it looks that it can be changed too and it looks that all callers use an existing object (so we can use a reference).

The only thing is that it makes the refactoring relatively large - but it is kept in a separate commit.

kbr-scylla commented on service/migration_manager.cc:
```diff
@@ -1094,7 +1095,10 @@ future<schema_ptr> migration_manager::get_schema_for_write(table_schema_version
         // Batch the barriers so we don't invoke them redundantly.
         mlogger.trace("Performing raft read barrier because schema is not synced, version: {}", v);
-        co_await _group0_barrier.trigger(as);
+        utils::composite_abort_source composite_as;
```
The abort sources are used inconsistently after this change.

I mean the `_as.abort_requested()` check at the beginning of this function; if we decide to compose the abort sources, shouldn't it be done at the beginning of the function, and all the checks refer to `composite_as`?

emdroid replied:
Yeah, that was in particular the eventual race condition I was referring to - only checked at the beginning, but not being checked later.

Looking through the history, the function was much simpler initially and the check at the beginning made sense, now it should probably be moved. I still kept the check, but moved it into the `get_schema_definition()` method that is being called from here (as we probably still want to skip that part if there is an abort request); which will in particular also be using the composite source.

emdroid commented:
New version:
* rebased on master
* `abort_source` pointer to reference refactoring in more places
* `get_schema_for_write()` abort source handling updates based on the review comments

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10620/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10620/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/group0_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10620/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2804/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10827/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10620/)
#### Failed Tests (25/31563):
- [test_repair_one_missing_row_diff_shard_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10620/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_one_missing_row_diff_shard_count) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_one_missing_row_diff_shard_count)
- [test_repair_one_missing_row_diff_shard_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10620/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_one_missing_row_diff_shard_count_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_one_missing_row_diff_shard_count)
- [test_repair_one_deleted_row_diff_shard_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10620/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_one_deleted_row_diff_shard_count) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_one_deleted_row_diff_shard_count)
- [test_repair_one_deleted_row_diff_shard_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10620/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_one_deleted_row_diff_shard_count_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_one_deleted_row_diff_shard_count)
- [test_repair_disjoint_row_2nodes_diff_shard_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10620/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_disjoint_row_2nodes_diff_shard_count) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_disjoint_row_2nodes_diff_shard_count)
- [test_repair_disjoint_row_2nodes_diff_shard_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10620/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_disjoint_row_2nodes_diff_shard_count_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_disjoint_row_2nodes_diff_shard_count)
- [test_repair_same_row_diff_value_3nodes_diff_shard_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10620/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_same_row_diff_value_3nodes_diff_shard_count) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_same_row_diff_value_3nodes_diff_shard_count)
- [test_repair_same_row_diff_value_3nodes_diff_shard_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10620/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_same_row_diff_value_3nodes_diff_shard_count_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_same_row_diff_value_3nodes_diff_shard_count)
- [test_repair_joint_row_3nodes_2_diff_shard_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10620/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_joint_row_3nodes_2_diff_shard_count) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_joint_row_3nodes_2_diff_shard_count)
- [test_repair_disjoint_row_3nodes_diff_shard_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10620/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_disjoint_row_3nodes_diff_shard_count) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_disjoint_row_3nodes_diff_shard_count)
#### Build Details:
- Duration: 5 hr 43 min
- Builder: i-0bf9368d84595baab (m5ad.8xlarge)

emdroid commented:
New version:
* rebased
* removed the last commit with `get_schema_for_write()` change - seems to break some dtests

The `get_schema_for_write()` change will need further investigation, therefore removed it for now (will create a separate follow-up ticket). It is not necessary to fix this particular issue (no tests pointed to it), so it can be done separately to not block this ticket.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10638/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10638/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/group0_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10638/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2820/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10841/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10638/)
#### Build Details:
- Duration: 2 hr 47 min
- Builder: spider8.cloudius-systems.com

kbr-scylla commented:
Removed last part of PR description during merge (no longer applies after the last commit was removed.)
